File: LaTeXDraw/src/net/sf/latexdraw/actions/TranslateShapes.java
package net.sf.latexdraw.actions;

import net.sf.latexdraw.glib.models.interfaces.GLibUtilities;
import net.sf.latexdraw.util.LNumber;

import org.malai.undo.Undoable;

/**
 * This action translates shapes.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2011 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 12/05/2010<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
public class TranslateShapes extends DrawingSelectionAction implements Undoable, Modifying {
	/** The x vector translation. */
	protected double tx;

	/** The y vector translation. */
	protected double ty;

	/** The x vector translation that has been already performed. This attribute is needed since
	 * this action can be executed several times. */
	private double performedTx;

	/** The y vector translation that has been already performed. This attribute is needed since
	 * this action can be executed several times. */
	private double performedTy;


	/**
	 * Initialises the action.
	 * @since 3.0
	 */
	public TranslateShapes() {
		super();

		tx = 0.;
		ty = 0.;
		performedTx = 0.;
		performedTy = 0.;
	}


	@Override
	public boolean isRegisterable() {
		return hadEffect();
	}


	@Override
	protected void doActionBody() {
		selection.translate(tx-performedTx, ty-performedTy);
		selection.setModified(true);
		drawing.setModified(true);
		performedTx = tx;
		performedTy = ty;
	}


	@Override
	public boolean canDo() {
		final boolean okShape = super.canDo() && !selection.isEmpty() && GLibUtilities.INSTANCE.isValidPoint(tx, ty);
		return okShape && (!LNumber.INSTANCE.equals(tx, 0.) || !LNumber.INSTANCE.equals(ty, 0.));
	}


	@Override
	public void undo() {
		selection.translate(-tx, -ty);
		selection.setModified(true);
		drawing.setModified(true);
	}


	@Override
	public void redo() {
		selection.translate(tx, ty);
		selection.setModified(true);
		drawing.setModified(true);
	}


	@Override
	public String getUndoName() {
		return "Translation";
	}


	/**
	 * @param tx The x vector translation.
	 * @since 3.0
	 */
	public void setTx(final double tx) {
		this.tx = tx;
	}


	/**
	 * @param ty The y vector translation.
	 * @since 3.0
	 */
	public void setTy(final double ty) {
		this.ty = ty;
	}
}

File: latexdraw-research/GUImutants/mutant45/net.sf.latexdraw.mutant45/src/main/net/sf/latexdraw/actions/shape/TranslateShapes.scala
package net.sf.latexdraw.actions.shape

import org.malai.action.Action
import org.malai.undo.Undoable

import net.sf.latexdraw.actions.DrawingAction
import net.sf.latexdraw.actions.Modifying
import net.sf.latexdraw.actions.ShapeAction
import net.sf.latexdraw.glib.models.interfaces.GLibUtilities
import net.sf.latexdraw.glib.models.interfaces.IGroup
import net.sf.latexdraw.util.LNumber

/**
 * This action translates shapes.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2013 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2012-04-20<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
class TranslateShapes extends Action with ShapeAction[IGroup] with DrawingAction with Undoable with Modifying {
	/** The x vector translation. */
	protected var _tx : Double = 0.0

	/** The y vector translation. */
	protected var _ty : Double = 0.0

	/** The x vector translation that has been already performed. This attribute is needed since
	 * this action can be executed several times. */
	private var performedTx : Double = 0.0

	/** The y vector translation that has been already performed. This attribute is needed since
	 * this action can be executed several times. */
	private var performedTy : Double = 0.0


	override def isRegisterable() = hadEffect


	protected def doActionBody() {
		_shape.get.translate(_tx-performedTx, _ty-performedTy)
		_shape.get.setModified(true)
		_drawing.get.setModified(true)
		//performedTx = _tx
		//performedTy = _ty
		//Mutant45
		performedTx = _ty
		performedTy = _tx
	}


	override def canDo() =  _drawing.isDefined && _shape.isDefined && !_shape.get.isEmpty && GLibUtilities.INSTANCE.isValidPoint(_tx, _ty)


	override def undo() {
		_shape.get.translate(-_tx, -_ty)
		_shape.get.setModified(true)
		_drawing.get.setModified(true)
	}


	override def redo() {
		_shape.get.translate(_tx, _ty)
		_shape.get.setModified(true)
		_drawing.get.setModified(true)
	}


	override def getUndoName() = "Translation"


	/**
	 * @param tx The x vector translation.
	 */
	def setTx(tx : Double) {
		_tx = tx
	}


	/**
	 * @param ty The y vector translation.
	 */
	def setTy(ty : Double) {
		_ty = ty
	}
}
